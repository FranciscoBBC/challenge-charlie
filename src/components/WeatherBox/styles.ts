import styled from 'styled-components';

export const BoxWrapper = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    width: 100%;
    max-width: 600px;
    height: 100%;
`;
