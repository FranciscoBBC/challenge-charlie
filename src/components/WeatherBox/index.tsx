import React, { useState, useEffect, useCallback } from 'react';
import CityInput from 'components/CityInput';
import WeatherStatus from 'components/WeatherStatus';

import GeoLocationService from 'services/GeoLocationService';
import CacheService from 'services/CacheService';
import WeatherService, { returnDailyWeather } from 'services/WeatherService';
import * as S from './styles';

export type locationProps = {
    lat: number;
    lng: number;
    city: string;
    state: string;
    stateCode: string;
    country: string;
};

const WeatherBox = () => {
    const [currentLocation, setCurrentLocation] = useState<locationProps>({
        lat: 0,
        lng: 0,
        city: '',
        state: '',
        stateCode: '',
        country: ''
    });
    const [currentWeather, setCurrentWeather] =
        useState<returnDailyWeather | null>(null);

    // utilizado para tratar o retorno de sucesso nos casos em que
    // foi possível pegar a localização do browser
    const handleSuccess = useCallback(
        async (pos: { coords: { latitude: number; longitude: number } }) => {
            try {
                const { coords } = pos;
                const locationInfos =
                    await GeoLocationService.getUserCityByLocation(
                        coords.latitude,
                        coords.longitude
                    );
                if (locationInfos) {
                    const location = {
                        lat: coords.latitude,
                        lng: coords.longitude,
                        city: locationInfos?.city,
                        state: locationInfos?.state,
                        stateCode: locationInfos?.state_code,
                        country: locationInfos?.country
                    };
                    CacheService.setCookie(
                        GeoLocationService.cacheLocationKey,
                        JSON.stringify(location),
                        5
                    );
                    setCurrentLocation(location);
                }
            } catch (error) {
                console.error(error);
            }
        },
        []
    );

    // utilizada para tratar exceções do serviço de geolocalização
    const handleError = useCallback(
        (err: { code: number; message: string }) => {
            // colocando como default as coordenadas default caso
            // usuário se recuse forcencer a localização
            if (err.code === 1) {
                handleSuccess({
                    coords: {
                        latitude: -22.91430764816798,
                        longitude: -47.058726209516394
                    }
                });
            } else {
                console.warn(
                    'ocorreu um erro ao solicitar a localização do usuário'
                );
                // tratamento de erro default
            }
        },
        [handleSuccess]
    );

    // função que ao iniciar o componente tenta pegar a posição atual do usuário
    const handlePosition = useCallback(() => {
        const localLocation = CacheService.getCookie(
            GeoLocationService.cacheLocationKey
        );
        if (localLocation.length <= 0) {
            GeoLocationService.getUserLocation(handleSuccess, handleError);
        } else {
            setCurrentLocation(JSON.parse(localLocation));
        }
    }, [handleError, handleSuccess]);

    // callback que retorna a localização completa do usuário baseada no que ele escreveu
    const handleChangeCity = useCallback((newLocation: locationProps) => {
        CacheService.setCookie(
            GeoLocationService.cacheLocationKey,
            JSON.stringify(newLocation),
            5
        );
        setCurrentLocation({ ...newLocation });
    }, []);

    const getWeatherInfos = useCallback(async () => {
        try {
            const { lat, lng } = currentLocation;
            if (lat && lng) {
                const response = await WeatherService.getWeatherNextDays(
                    lat,
                    lng
                );
                if (response) {
                    setCurrentWeather(response);
                }
            }
        } catch (error) {
            console.error(error);
        }
    }, [currentLocation]);

    useEffect(() => {
        handlePosition();
    }, []);

    useEffect(() => {
        getWeatherInfos();
    }, [getWeatherInfos]);

    return (
        <S.BoxWrapper>
            <CityInput
                state={currentLocation?.state}
                city={currentLocation?.city}
                onChangeCity={handleChangeCity}
            />
            <WeatherStatus dailyWeather={currentWeather} />
        </S.BoxWrapper>
    );
};

export default WeatherBox;
