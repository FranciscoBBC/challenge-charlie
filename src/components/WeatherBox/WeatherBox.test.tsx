import React from 'react';
import { render } from '@testing-library/react';
import WeatherBox from '.';

describe('<WeatherBox />', () => {
    it('should be rendered', () => {
        const { container } = render(<WeatherBox />);

        expect(container.firstChild).toBeTruthy();

        expect(container.firstChild).toMatchSnapshot();
    });
});
