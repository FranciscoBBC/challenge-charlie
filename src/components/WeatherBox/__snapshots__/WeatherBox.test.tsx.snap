// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherBox /> should be rendered 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background: #fff;
  border: none;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  height: 50px;
  opacity: 85%;
  box-shadow: 0 1px 5px #888888;
  z-index: 1;
}

.c2 {
  padding: 0 5px 0 5px;
}

.c3 {
  width: 100%;
}

.c4 {
  width: 90%;
  border: none;
  font-size: 26px;
  font-weight: bold;
  color: #8d8986;
}

.c4:focus {
  outline: none;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 8;
  -ms-flex: 8;
  flex: 8;
  width: 100%;
  height: auto;
}

.c6 {
  display: grid;
  grid-template-columns: 4fr 3fr;
  grid-template-rows: 3fr 1fr 1fr;
  width: 100%;
  height: 100%;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  opacity: 85%;
  height: 100%;
  background: #606060;
  padding: 0 5px 0 5px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  background: #3D3D3D;
  opacity: 93%;
  padding: 0 5px 0 5px;
}

.c18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  background: #262626;
  padding: 0 5px 0 5px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  background: #606060;
  opacity: 85%;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  background: #3D3D3D;
  opacity: 93%;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  background: #262626;
  opacity: 1;
}

.c9 {
  cursor: auto;
  color: #fff;
  font-size: 1.8rem;
  font-weight: 400;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  padding: 8px 0 3px 0;
}

.c10 {
  cursor: pointer;
  color: #fff;
  font-size: 1.8rem;
  font-weight: 400;
  font-family: 'Open Sans';
  padding: 0;
}

.c11 {
  cursor: auto;
  color: #fff;
  font-size: 2rem;
  font-weight: 400;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  padding: 16px 0 10px 0;
}

.c12 {
  cursor: auto;
  color: #fff;
  font-size: 1.4rem;
  font-weight: 400;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  padding: 0;
}

.c15 {
  cursor: auto;
  color: #fff;
  font-size: 1.5rem;
  font-weight: 400;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  padding: 8px 0 3px 0;
}

.c16 {
  cursor: auto;
  color: #fff;
  font-size: 1.5rem;
  font-weight: 400;
  font-family: 'Open Sans';
  padding: 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  max-width: 600px;
  height: 100%;
}

@media screen and (max-width:768px) {
  .c8 p:first-child {
    padding: 0;
  }
}

@media screen and (max-width:425px) {
  .c14 p {
    font-size: 18px;
  }
}

@media screen and (max-width:425px) {
  .c18 p {
    font-size: 18px;
  }
}

@media screen and (max-width:768px) {
  .c9 {
    font-size: 1.5rem;
    font-weight: 400;
  }
}

@media screen and (max-width:768px) {
  .c10 {
    font-size: 1.5rem;
    font-weight: 400;
  }
}

@media screen and (max-width:768px) {
  .c11 {
    font-size: 1.8rem;
    font-weight: 400;
  }
}

@media screen and (max-width:768px) {
  .c12 {
    font-size: 1.2rem;
    font-weight: 400;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="city-input-holder"
  >
    <div
      class="c2"
    >
      <svg
        color="8d8986"
        fill="currentColor"
        height="50px"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="50px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
        />
        <path
          d="m8 16 5.991-2L16 8l-6 2z"
        />
      </svg>
    </div>
    <div
      class="c3"
    >
      <input
        class="c4"
        data-testid="city-input"
        placeholder=""
        type="text"
      />
    </div>
  </div>
  <div
    class="c5"
  >
    <div
      class="c6"
    >
      <div
        class="c7"
        data-testid="weather-icon"
        opacity="85%"
      />
      <div
        class="c8"
        data-testid="today-weather"
      >
        <p
          class="c9"
          font-size="1.8rem"
        >
          HOJE
        </p>
        <p
          class="c10"
          cursor="pointer"
          data-testid="temperature"
          font-family="'Open Sans'"
          font-size="1.8rem"
        >
          0ºC
        </p>
        <p
          class="c11"
          font-size="2rem"
        />
        <p
          class="c12"
          data-testid="wind"
          font-size="1.4rem"
        >
          Vento: 
          
        </p>
        <p
          class="c12"
          data-testid="humidity"
          font-size="1.4rem"
        >
          Humidade: 
          %
        </p>
        <p
          class="c12"
          data-testid="pressure"
          font-size="1.4rem"
        >
          Pressão: 
          hPA
        </p>
      </div>
      <div
        class="c13"
        opacity="93%"
      />
      <div
        class="c14"
        data-testid="tomorrow-weather"
      >
        <p
          class="c15"
          font-size="1.5rem"
        >
          AMANHÃ
        </p>
        <p
          class="c16"
          font-family="'Open Sans'"
          font-size="1.5rem"
        >
          0ºC
        </p>
      </div>
      <div
        class="c17"
      />
      <div
        class="c18"
        data-testid="aftertomorrow-weather"
      >
        <p
          class="c15"
          font-size="1.5rem"
        >
          DEPOIS DE AMANHÃ
        </p>
        <p
          class="c16"
          font-family="'Open Sans'"
          font-size="1.5rem"
        >
          0ºC
        </p>
      </div>
    </div>
  </div>
</div>
`;
